### PR TITLE
cephfs: add support for "snapshot-autoprotect" feature

### PR DESCRIPTION
**Pull request by Yuggupta27**
Subvolume snapshots required to be protected, prior to
cloning the same. Ceph v14.2.12 brought in the feature
of "snapshot-autoprotect".
Adding support for the same in ceph-csi.

Signed-off-by: Yug Gupta <yuggupta27@gmail.com>

Fixes: #1338

### CODE REVIEW

File: internal/cephfs/clone.go
Comment (Madhu-1) [inline internal/cephfs/clone.go:68]: to keep the code changes minimal, as the `volOptions` are passed to both protect and unprotect function how about checking `checkSubvolumeHasFeature` inside `protect` and `unprotect` function?

**Comment by Yuggupta27** [comment]
Since the e2e passed for the current ci, added a temporary commit to verify changes with ceph latest-nautilus-devel.
cc @Madhu-1 @humblec

File: internal/cephfs/clone.go
Comment (Yuggupta27) [reply internal/cephfs/clone.go:69]: Checking outside might prevent many extra calls to `protectSnapshot` and `unprotectSnapshot`. Thoughts?

Comment (Madhu-1) [reply internal/cephfs/clone.go:69]: Checking inside will check for the feature on subvolume if the feature is present the call will be treated as a no-op

Comment (Yuggupta27) [reply internal/cephfs/clone.go:69]: Thanks for the input. I'll make the changes.

File: internal/cephfs/snapshot.go
Comment (Madhu-1) [inline internal/cephfs/snapshot.go:140]:
```
if checkSubvolumeHasFeature("snapshot-autoprotect", volOptions.Features) {
  return nil
}
```

File: internal/cephfs/snapshot.go
Comment (Madhu-1) [inline internal/cephfs/snapshot.go:173]:
```
if checkSubvolumeHasFeature("snapshot-autoprotect", volOptions.Features) {
  return nil
}
```

File: internal/cephfs/snapshot.go
Comment (Yuggupta27) [reply internal/cephfs/snapshot.go:140]: I didn't follow this approach with an idea to keep the `return` statements minimum. The current code also looks to be doing the same thing to me. If this is a better approach, I would love to know the reason :)

Comment (Madhu-1) [reply internal/cephfs/snapshot.go:140]: with this approach, the number of changed lines of code is 3+3(which makes review easier) lines for protect and unprotect isn't it?

Comment (Yuggupta27) [reply internal/cephfs/snapshot.go:140]: > with this approach, the number of changed lines of code is 3+3(which makes review easier) lines for protect and unprotect isn't it?

Thanks for explaining :) I'll change it accordingly.

**Comment by Yuggupta27** [comment]
@Madhu-1 , Should I remove the temporary commit now?

**Comment by Madhu-1** [comment]
@Yuggupta27 Hope you have tested and there was no protect and unprotect call issued from cephcsi side during snapshot create and pvc-pvc clone testing?

**Comment by nixpanic** [comment]
Where does the string `snapshot-autoprotect` come from (in the Ceph sources)? Could you verify with some ceph commands that this is set on new Ceph deployments? Please copy/paste the commands and output that you used for verification.

**Comment by Yuggupta27** [comment]
> Where does the string `snapshot-autoprotect` come from (in the Ceph sources)? Could you verify with some ceph commands that this is set on new Ceph deployments? Please copy/paste the commands and output that you used for verification.

Sure
```
sh-4.4# ceph version      
ceph version 14.2.11-132-g36fefe6797 (36fefe6797b96c76b540931bc7aac84335ac34d0) nautilus (stable)
```
```
sh-4.4# ceph fs subvolume create  myfs myvol
sh-4.4# ceph fs subvolume info myfs myvol
{
    "atime": "2020-09-15 11:57:55", 
    "bytes_pcent": "undefined", 
    "bytes_quota": "infinite", 
    "bytes_used": 0, 
    "created_at": "2020-09-15 11:57:55", 
    "ctime": "2020-09-15 11:57:55", 
    "data_pool": "myfs-data0", 
    "features": [
        "snapshot-clone", 
        "snapshot-autoprotect", 
        "snapshot-retention"
    ], 
    "gid": 0, 
    "mode": 16877, 
    "mon_addrs": [
        "10.96.74.216:6789"
    ], 
    "mtime": "2020-09-15 11:57:55", 
    "path": "/volumes/_nogroup/myvol/70beb26e-5fe1-424e-bda3-0e7d4f19fb1f", 
    "pool_namespace": "", 
    "state": "complete", 
    "type": "subvolume", 
    "uid": 0
}
```

**Review by nixpanic — APPROVED** [review]

**Comment by humblec** [comment]
LGTM.. However holding approval for @Madhu-1 to take a look at this, as he has reviewed this PR.

**Comment by Madhu-1** [comment]
> @Yuggupta27 Hope you have tested and there was no protect and unprotect call issued from cephcsi side during snapshot create and pvc-pvc clone testing?

waiting for a response from @Yuggupta27

**Comment by Yuggupta27** [comment]
@Madhu-1 , Here are the plugin logs in case the "auto-protect" feature is present on the subvolume:
```
I0915 13:03:52.085348       1 utils.go:165] ID: 32 GRPC response: {"name":"cephfs.csi.ceph.com","vendor_version":"protec"}
I0915 13:03:52.104885       1 utils.go:159] ID: 33 Req-ID: snapshot-4bafb874-9885-4db9-ac54-4249cdecd8a1 GRPC call: /csi.v1.Controller/CreateSnapshot
I0915 13:03:52.111714       1 utils.go:160] ID: 33 Req-ID: snapshot-4bafb874-9885-4db9-ac54-4249cdecd8a1 GRPC request: {"name":"snapshot-4bafb874-9885-4db9-ac54-4249cdecd8a1","parameters":{"clusterID":"myclusterID"},"secrets":"***stripped***","source_volume_id":"0001-000b-myclusterID-0000000000000001-7f87c127-f753-11ea-9425-0242ac110013"}
I0915 13:03:52.658692       1 cephcmds.go:59] ID: 33 Req-ID: snapshot-4bafb874-9885-4db9-ac54-4249cdecd8a1 command succeeded: ceph [-m 10.96.74.216:6789 --id admin --keyfile=***stripped*** -c /etc/ceph/ceph.conf fs dump --format=json]
I0915 13:03:53.042930       1 cephcmds.go:59] ID: 33 Req-ID: snapshot-4bafb874-9885-4db9-ac54-4249cdecd8a1 command succeeded: ceph [-m 10.96.74.216:6789 --id admin --keyfile=***stripped*** -c /etc/ceph/ceph.conf fs ls --format=json]
I0915 13:03:53.049770       1 omap.go:72] ID: 33 Req-ID: snapshot-4bafb874-9885-4db9-ac54-4249cdecd8a1 got omap values: (pool="myfs-metadata", namespace="csi", name="csi.volume.7f87c127-f753-11ea-9425-0242ac110013"): map[csi.imagename:csi-vol-7f87c127-f753-11ea-9425-0242ac110013 csi.volname:pvc-1ba6ae9e-e1ec-4758-8984-ad1f56228c3d]
I0915 13:03:54.209775       1 cephcmds.go:59] ID: 33 Req-ID: snapshot-4bafb874-9885-4db9-ac54-4249cdecd8a1 command succeeded: ceph [fs subvolume info myfs csi-vol-7f87c127-f753-11ea-9425-0242ac110013 --group_name mysubgrp -m 10.96.74.216:6789 -c /etc/ceph/ceph.conf -n client.admin --keyfile=***stripped***]
I0915 13:03:54.211291       1 omap.go:72] ID: 33 Req-ID: snapshot-4bafb874-9885-4db9-ac54-4249cdecd8a1 got omap values: (pool="myfs-metadata", namespace="csi", name="csi.snaps.default"): map[]
I0915 13:03:55.485959       1 cephcmds.go:59] ID: 33 Req-ID: snapshot-4bafb874-9885-4db9-ac54-4249cdecd8a1 command succeeded: ceph [fs subvolume info myfs csi-vol-7f87c127-f753-11ea-9425-0242ac110013 --group_name mysubgrp -m 10.96.74.216:6789 -c /etc/ceph/ceph.conf -n client.admin --keyfile=***stripped***]
I0915 13:03:55.516708       1 omap.go:136] ID: 33 Req-ID: snapshot-4bafb874-9885-4db9-ac54-4249cdecd8a1 set omap keys (pool="myfs-metadata", namespace="csi", name="csi.snaps.default"): map[csi.snap.snapshot-4bafb874-9885-4db9-ac54-4249cdecd8a1:e930db6a-f753-11ea-9425-0242ac110013])
I0915 13:03:55.522453       1 omap.go:136] ID: 33 Req-ID: snapshot-4bafb874-9885-4db9-ac54-4249cdecd8a1 set omap keys (pool="myfs-metadata", namespace="csi", name="csi.snap.e930db6a-f753-11ea-9425-0242ac110013"): map[csi.imagename:csi-snap-e930db6a-f753-11ea-9425-0242ac110013 csi.snapname:snapshot-4bafb874-9885-4db9-ac54-4249cdecd8a1 csi.source:csi-vol-7f87c127-f753-11ea-9425-0242ac110013])
I0915 13:03:55.522530       1 fsjournal.go:283] ID: 33 Req-ID: snapshot-4bafb874-9885-4db9-ac54-4249cdecd8a1 Generated Snapshot ID (0001-000b-myclusterID-0000000000000001-e930db6a-f753-11ea-9425-0242ac110013) for request name (snapshot-4bafb874-9885-4db9-ac54-4249cdecd8a1)
I0915 13:03:56.413007       1 cephcmds.go:59] ID: 33 Req-ID: snapshot-4bafb874-9885-4db9-ac54-4249cdecd8a1 command succeeded: ceph [fs subvolume snapshot create myfs csi-vol-7f87c127-f753-11ea-9425-0242ac110013 csi-snap-e930db6a-f753-11ea-9425-0242ac110013 --group_name mysubgrp -m 10.96.74.216:6789 -c /etc/ceph/ceph.conf -n client.admin --keyfile=***stripped***]
I0915 13:03:56.988404       1 cephcmds.go:59] ID: 33 Req-ID: snapshot-4bafb874-9885-4db9-ac54-4249cdecd8a1 command succeeded: ceph [fs subvolume snapshot info myfs csi-vol-7f87c127-f753-11ea-9425-0242ac110013 csi-snap-e930db6a-f753-11ea-9425-0242ac110013 --group_name mysubgrp -m 10.96.74.216:6789 -c /etc/ceph/ceph.conf -n client.admin --keyfile=***stripped*** --format=json]
I0915 13:03:56.990204       1 utils.go:165] ID: 33 Req-ID: snapshot-4bafb874-9885-4db9-ac54-4249cdecd8a1 GRPC response: {"snapshot":{"creation_time":{"nanos":358928000,"seconds":1600175036},"ready_to_use":true,"size_bytes":1073741824,"snapshot_id":"0001-000b-myclusterID-0000000000000001-e930db6a-f753-11ea-9425-0242ac110013","source_volume_id":"0001-000b-myclusterID-0000000000000001-7f87c127-f753-11ea-9425-0242ac110013"}}
```

**Review by Madhu-1 — COMMENTED** [review]
In some cases we do check snapshot info of the snapshot and do unprotect or protect based on the `protected` field as with current implementation both `protect and `unprotect` will become no-op it should be fine. 

LGTM from my side, @ShyamsundarR PTAL

File: internal/cephfs/snapshot.go
Comment (Madhu-1) [inline internal/cephfs/snapshot.go:142]: as `snapshot-autoprotect` is used in 2 places can you define it as const above and use it?

Comment (Yuggupta27) [reply internal/cephfs/snapshot.go]: Done.

File: internal/cephfs/snapshot.go
Comment (ShyamsundarR) [inline internal/cephfs/snapshot.go:182]: `volumeOptions` missing Features when invoked from `DeleteSnapshot`.

`volOpt` in DeleteSnapshot is populated in function `newSnapshotOptionsFromID` which does not invoke parent vol infor to read features and populate the same.

Comment (Madhu-1) [reply internal/cephfs/snapshot.go:182]: @ShyamsundarR as subvolume and snapshots are independent is it possible to get subvolume info if subvolume is already deleted?

Comment (ShyamsundarR) [reply internal/cephfs/snapshot.go:182]: Subvolume when deleted using `--retain-snapshots` retains the subvolume in a retained state, and `subvolume info` is still valid and will return the subvolume `type`, `features`, and `state` to the caller.

File: internal/cephfs/volumeoptions.go
Comment (Madhu-1) [inline internal/cephfs/volumeoptions.go:482]: we dont need this check. snapshot creation is not supported if the `subvolume info` command is not supported just return errror

File: internal/cephfs/snapshot.go
Comment (Madhu-1) [inline internal/cephfs/snapshot.go:30]:
```suggestion
const (
           autoProtect = "snapshot-autoprotect"
          )
```
lets add `()` so that we can add more consts later

File: internal/cephfs/volumeoptions.go
Comment (Yuggupta27) [reply internal/cephfs/volumeoptions.go:482]: Okay, This will not break any backward compatibility. Right?

File: internal/cephfs/snapshot.go
Comment (Yuggupta27) [reply internal/cephfs/snapshot.go:30]: Thanks, Will change it.

File: internal/cephfs/volumeoptions.go
Comment (Madhu-1) [reply internal/cephfs/volumeoptions.go:482]: Yes it should work

**Comment by Madhu-1** [comment]
>go: k8s.io/kubernetes@v1.18.6 requires
	github.com/caddyserver/caddy@v1.0.3 requires
	github.com/klauspost/cpuid@v1.2.0: reading github.com/klauspost/cpuid/go.mod at revision v1.2.0: unknown revision v1.2.0
ERROR: go.mod was modified by 'go mod verify'

go mod verify is failing @Yuggupta27 can you send a PR to fix it?

**Comment by Yuggupta27** [comment]
> > go: k8s.io/kubernetes@v1.18.6 requires
> > github.com/caddyserver/caddy@v1.0.3 requires
> > github.com/klauspost/cpuid@v1.2.0: reading github.com/klauspost/cpuid/go.mod at revision v1.2.0: unknown revision v1.2.0
> > ERROR: go.mod was modified by 'go mod verify'
> 
> go mod verify is failing @Yuggupta27 can you send a PR to fix it?

Yes Madhu, Will send a PR for that one.

**Review by ShyamsundarR — COMMENTED** [review]
LGTM! Will wait for other comments to be sorted...

**Review by Madhu-1 — APPROVED** [review]
LGTM

**Comment by Madhu-1** [comment]
@ShyamsundarR can you approve the PR if everything looks good?

**Review by ShyamsundarR — APPROVED** [review]

**Comment by Madhu-1** [comment]
@mergifyio backport release-v3.1

**Pull request merged**
